### ISSUE TITLE
[🐞] Actual value for useContext() can not be found, when using a component, that dynamically shows children using <Slot />

### ISSUE THREAD

**Issue by levno-710** [QwikDev/qwik#3531]
### Which component is affected?

Qwik Runtime

### Describe the bug

I am trying to create a component, that shows a loading screen, while fetching some data, and then shows it's components using <Slot />. The fetched data should be made available by useContext.
However, this throws an error while the components are not Shown.

Expected behaviour: It should work without error.

A workaround for this is, to move the useContextProvider to another component, that has the one that decides wether to render the children as a child.
[Here](https://stackblitz.com/edit/qwik-starter-o589ex?file=src/routes/index.tsx) is the non working example fixed using this workaround.

### Reproduction

https://stackblitz.com/edit/qwik-starter-6pi4u4?file=src/routes/index.tsx

### Steps to reproduce

Run `npm install && npm start`

### System Info

```shell
System:
    OS: Windows 10 10.0.19044
    CPU: (24) x64 AMD Ryzen 9 5900X 12-Core Processor
    Memory: 50.63 GB / 63.91 GB
  Binaries:
    Node: 19.2.0 - C:\Program Files\nodejs\node.EXE
    Yarn: 1.22.18 - ~\AppData\Roaming\npm\yarn.CMD
    npm: 8.19.3 - C:\Program Files\nodejs\npm.CMD
  Browsers:
    Chrome: 111.0.5563.111
    Edge: Spartan (44.19041.1266.0), Chromium (111.0.1661.54)
    Internet Explorer: 11.0.19041.1566
  npmPackages:
    @builder.io/qwik: 0.23.0 => 0.23.0
    @builder.io/qwik-city: 0.6.5 => 0.6.5
    undici: 5.21.0 => 5.21.0
    vite: 4.2.0 => 4.2.0
```

### Additional Information

The example is under the index.tsx route.

**Comment by peter-kuhmann** [comment]
I am experiencing the same bug. I experienced it while creating a multi-tenant application and using the context API for sharing the tenant information.

**Comment by peter-kuhmann** [comment]
I found the problem in my case. Here is what I initially tried:

- `routes/layout.tsx`
    - Fetch data using a `routerLoader$`
    - Expose loaded data via `useContextProvider`
    - Render children conditionally using `{condition && <Slot />}`
- `routes/index.tsx`
    - Get access to data via `useContext`    

In my case, the problem was:
- Qwik City seems to render first the route component (so `index.tsx`) and then _projects_ that component into the Slot of `layout.tsx`.
- This order (first `index.tsx`, then `layout.tsx`) prevents `layout.tsx` from reliably setting context via `useContextProvider`.

It turns out that the simplest fix was **ditching the context**. I could just use the router loader directly in `index.tsx` 🤦 😅 

@levno-710 Maybe this helps you as well?

**Comment by icflorescu** [comment]
> Experienced the same bug. I was hoping to use this to help fill the void that the lack of render props/slot props leaves behind. I'm currently getting around this by using a function prop (ie. `<Menu.Item render$={({ active }) => <MyComp active={active} />} />`

@NickClark does this work in production mode for you (with `qwik build preview && vite preview --open`)?

**Comment by NickClark** [comment]
@icflorescu Yes, it appears to.

**Comment by NickClark** [comment]
So, it appears that a child that is slotted in, even if not rendered by the parent, will still get rendered on the server. So if the parent renders `null`, it isn't mounted, but the child still get's rendered, but since there's technically no parent, it doesn't get the context either. This seems wrong. 

For instance, in this example, I would never expect the body of `Child` to render, but it does.

```tsx
import { component$ } from "@builder.io/qwik";

export default component$(() => {
  return (
    <>
      <Parent>
        <Child>
          Child content
        </Child>
      </Parent>
    </>
  );
});

export const Parent = component$(() => {
    return null
});

export const Child = component$(() => {
  throw new Error("Child should not be rendered");
});

```

**Comment by NickClark** [comment]
Consider also the following:

```tsx

export const TestContextId = createContextId<Signal<boolean>>("TestContext");

export default component$(() => {
  return (<Parent> <Child /> </Parent>);
});

export const Parent = component$(() => {
  const isOpen = useSignal(false);
  useContextProvider(TestContextId, isOpen);

  return (
    <>
      <button onClick$={() => (isOpen.value = !isOpen.value)}>Test - {isOpen.value ? "true" : "false"}</button>
      <div>
        {isOpen.value && < Slot />}
      </div>
    </>
  );
});

export const Child = component$(() => {
  const isOpen = useContext(TestContextId, null);
  console.log("Child isOpen", isOpen?.value);
  return (
      <span>Test - `{JSON.stringify(isOpen, null, 2)}` - `{isOpen?.value ? "true" : "false"}`</span>
  );
});
```

If the signal defaults to true, everything will work correctly. But if it starts `false`, then the context isn't successfully setup, and the child will never receive `isOpen` signals, since on initialization on the server, the signal didn't exist. This is a pretty huge bug for us... I'm quite surprised more developers haven't run hard into this.

This has also made me wonder... even if a child template isn't being used... do all dependents have to get rendered on the server? In this case it seems like the child shouldn't have been included in the initial render at all... it should have been progressively loaded on click.

I normally don't ping people directly, but... @manucorporat ... do you have any insight on this for us? This seems like a pretty big breaking bug for contexts. Maybe I'm just missing something.

**Comment by gparlakov** [comment]
Hi, having the same issue with a different spin to it:

Here's a repro on stackblitz: 
https://stackblitz.com/edit/qwik-starter-b6shxk?file=src%2Froutes%2Findex.tsx 

(there's a working example in /working that seems to do the same but works ?)


The context:

- building a wizard component 
- would like to switch the step based on a `current` step and showing the <Slot /> for that step only 
- there are crumbs for the wizard that show all slots up to the current step 
  - so we might see <Slot name="step-2"/> and <Slot name="crumb-0" /><Slot name="crumb-2" /><Slot name="crumb-2" />

**Comment by RaeesBhatti** [comment]
This would be really useful to resolve user information once at the top-level layout and access it wherever needed.

**Comment by wmertens** [comment]
See also #5270 which might be the same issue. I'm looking into it.

**Comment by gioboa** [comment]
This is fixed in #5263
